### DISCUSSION TITLE
(another) glb load error on iOS

### DISCUSSION THREAD

**Discussion by clochardM33** [google/model-viewer#3564] [Q&A] [answered]
### Description

Hello. I have a model file that crashes the web page (only) on iOS.
It seems to be related to one specific material as when that material is deleted or swapped out the page loads properly.
The issue looks a lot like #3000 and #3067 , but it post dates the fixes in 1.12 for those two, hence this entry.

The fault manifests itself as:
Model loading starts.
Progress bar gets to about 1/3 
Very briefly there is bit of text on the page, something about error loading file
Model-viewer then "resets" in that the poster is refreshed and the loading starts again, but this time it skips to the end and then chucks a page error along the lines of "A problem repeatedly occurred on..."

Apart from experimenting with the content of the glb I don't have any other capability to diagnose anything sorry.

The fault occurs on both Safari and (iOS) Firefox. Although with Firefox it drops into a continual reload loop rather than stopping with an error message.

Attached are two glb: the one that crashes and the same file with "Material 26" removed (which is the one that loads correctly).
"Material 26" was generated from a png image mixed with a base colour.

The complication is that the model was created in all innocence, with a very normal workflow and only when it got to the iOS configuration did it not work. 

And the final twist is that the glitch site below (which is one page with both models) works fine, whereas the same page either on localhost or on our internal server gives the error.  So now I'm really confused.

#### Live Demo

https://chestnut-lapis-utahraptor.glitch.me

### Version

- model-viewer: v1.12 and v1.8

### Browser Affected

- [ ] Chrome, version: xx.x.xxxx.xx
- [ ] Edge
- [ 1] Firefox
- [ ] IE
- [1 ] Safari

### OS

- [ ] Android
- [ 1 ] iOS v14 and v15.5
- [ ] Linux
- [ ] MacOS
- [ ] Windows

[Issue_Texture_ios.zip](https://github.com/google/model-viewer/files/9018485/Issue_Texture_ios.zip)

**Comment by elalish** [comment, marked as answer]
So the issue doesn't reproduce on your Glitch? That sounds more like a server configuration problem. Not sure I can do much to help here...

**Reply by clochardM33** [reply]
Update: The issue _does_ occur on the Glitch link. 
Manged to get some time on 3 different iPhones all running iOS 15.5 and it is a consistent fault.
The ancient iPhone SE on iOS 14 that is my only test gear shows different behaviour (hence the final line above). This loads, but takes does a reload (as described above), but if you reload the page then it folds in the way the other phones do. If you open a new tab then you get the first time lucky behaviour.
It has to be something about the structure of the models, but I'm at a loss to work out what it is. Any guidance or insight would be appreciated, but understand that this might not be to do with m-v. Thanks.

**Reply by elalish** [reply]
Strange, when I try to download your GLBs chrome says they are "blocked", maybe a weird CORS setting? Also they're 15Mb, which is pretty large; you might just be hitting some texture size or GPU RAM limits.

**Reply by clochardM33** [reply]
Thanks for looking at this. Why is it never straightforward?
If you are downloading from the Glitch URL then you will have to temporarily switch off Safe Browsing as that blocks the download (I'm guessing that's it).
https://support.google.com/chrome/answer/9890866
The same files are also an attachment in the top message as the zip file.

It isn't the size that is directly the problem as for our model sets this is small.
The two files mentioned above have a runtime useage of 16Mb for the "good" one and 9Mb for the "bad" one (I forgot to compress the "good" one).
We have a whole range of sizes and m-v runs beautifully on all of them (on a PC), but the biggest ones do choke older phones.
Most models are in the range 50Mb to 200Mb and an old iPhone SE can just hold that together (you can forget AR at that size though). The largest we have is 325Mb (runtime) and that throws an error in everything but the newest phones.
An obersvation is that on Safari (on iOS) for this massive model it attempts a reload, which is what is happening with the failing file (the one at the start of this thread).

**Reply by elalish** [reply]
Well, I can repro the reload and eventual crash on my old iPhone SE on iOS 15. Frankly I shouldn't be able to crash a browser even if I try, so this looks like a WebKit bug to me. Would you mind filing it on them? They had a number of WebGL regressions with iOS 15. @grorg in case Apple has any insight here.

**Reply by clochardM33** [reply]
By all means send it on. Good to know that I'm not goind mad. Thanks.

**Comment by henrik-dons** [comment]
Hello all,

I have issue with the GLB 3D model on iOS devices.
Demo page: https://divatude.co/product/tbs/

Model does not load on iOS Safari and Chrome.
It uses PBR textures so they can be changed dynamically on the page.
GLB model is valid according to Khronos Validator (https://github.khronos.org/glTF-Validator/)

Any ideas what could be causing this issue?

Best regards,
Henrik

**Reply by clochardM33** [reply]
It has all the hallmarks of running out of resources. See the GPU overlay on these two Chrome screenshot. The FPS chugs below 20 and everything is trying hard. 
This is on a few year old Galaxy S10e. 
So it isn't inherently broken. 
And it looks lovely in my back yard, but I couldn't afford a pair. 


![Screenshot_20240621_162601_Chrome](https://github.com/google/model-viewer/assets/88084694/f5adf0f2-f70b-425d-8187-f8047a2e5c4a)
![Screenshot_20240621_162528_Chrome](https://github.com/google/model-viewer/assets/88084694/9672fcd9-2f04-43f4-a4e2-eb6be615ac76)

**Reply by clochardM33** [reply]
Hi Henrik,
I found a bit of time to look at your model, hope you don't mind.
It has lots of verts, lots of them.  If your non-load isn't something technical then it's just the triangle count.
The glitch below has a lower poly version of your glb that I worked up. On a phone it loads in about 60% of the time (load is when you can interact with it). 
(The product variants are spread to the side to make the number of verts a little more obvious. This isn't how your model was arranged.)

Pull the model out into Blender and it needs ~9Gb of RAM to hold it!
Before there were about 5.7 million verts, after 3.9 million. And most of them you can't see as the object is so small in real scale.
The heavy numbers come from the circular wire in the gem holders, but that is a bit unavoidable, and probably in the triangulation getting the objects from CAD into glb-land.

Hope this helps.

https://cerulean-glorious-hero.glitch.me

[Gem Shoe - Sheet1.csv](https://github.com/user-attachments/files/15958299/Gem.Shoe.-.Sheet1.csv)

**Reply by henrik-dons** [reply]
Hello Clochard,
Thank you so much for looking into this and sharing an example!
I will share this with the 3D designer, hope he gets the idea :)
Cheers!

**Reply by henrik-dons** [reply]
Hello guys,

Any idea why the model viewer on this page crashes?
Works fine on desktop, but crashes on mobile devices...
Can't seem to find any issues with the model itself.

https://pedocpower.com/product/product-to-test-errors/

Any help or insights would be much appreciated!

**Reply by elalish** [reply]
It's working fine on my Pixel 7 - what mobile devices is it crashing on? Also, for you side-scrolling unit you might consider `touch-action="pan-x"`: https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-touch-action

**Reply by henrik-dons** [reply]
Hello, sorry it works fine on Android device but not on iPhone.
Tested with iPhone 12 on Chrome and Safari, both crash on page load.

**Reply by elalish** [reply]
Chrome and Safari are the same on iOS - Apple only allows WebKit browsers. The page certainly shouldn't crash, even if we do something terrible. Please file an issue on WebKit.
